geode-core/src/distributedTest/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceConstraintsDistributedTest.java

```diff
@@ -0,0 +1,595 @@
+      assertThat(result.size()).isEqualTo(ENTRIES);
+    });
+
+    // Execute query again, it should fail as the restricted authorizer has been installed.
+    server.invoke(QueryObserverHolder::reset);
+    client.invoke(() -> {
+      assertThat(ClusterStartupRule.getClientCache()).isNotNull();
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      Query newQuery = clientCache.getQueryService().newQuery(queryString);
+      assertThatThrownBy(newQuery::execute)
+          .isInstanceOf(ServerOperationException.class)
+          .hasCauseInstanceOf(NotAuthorizedException.class)
+          .hasStackTraceContaining(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "getName");
+    });
+  }
+
+  /**
+   * The test registers a {@link QueryObserver} that changes the installed
+   * {@link MethodInvocationAuthorizer} to the default one (denies everything) as soon as the query
+   * starts, and makes sure the query succeeds (it is no affected by the authorizer change).
+   * The query is executed from the server.
+   */
+  @Test
+  @Parameters({"LOCAL", "REPLICATE", "PARTITION"})
+  @TestCaseName("[{index}] {method}(RegionType:{0})")
+  public void queriesInFlightExecutedByServersShouldNotBeAffectedWhenMethodAuthorizerIsChanged(
+      RegionShortcut regionShortcut) {
+    String regionName = testName.getMethodName();
+    createAndPopulateRegion(regionName, regionShortcut);
+    String queryString = "<TRACE> SELECT object.getName() FROM /" + regionName + " object";
+
+    server.invoke(() -> {
+      // Set test query observer.
+      assertThat(ClusterStartupRule.getCache()).isNotNull();
+      InternalCache internalCache = ClusterStartupRule.getCache();
+      TestQueryObserver queryObserver =
+          new TestQueryObserver(RestrictedMethodAuthorizer.class.getName(), Collections.emptySet());
+      QueryObserverHolder.setInstance(queryObserver);
+
+      // Execute query.
+      Query query = internalCache.getQueryService().newQuery(queryString);
+      @SuppressWarnings("unchecked")
+      SelectResults<QueryObject> result = (SelectResults<QueryObject>) query.execute();
+      assertThat(queryObserver.invocations.get()).isEqualTo(1);
+      assertThat(result.size()).isEqualTo(ENTRIES);
+
+      // Execute query again, it should fail as the restricted authorizer has been installed.
+      QueryObserverHolder.reset();
+      Query newQuery = internalCache.getQueryService().newQuery(queryString);
+      assertThatThrownBy(newQuery::execute)
+          .isInstanceOf(NotAuthorizedException.class)
+          .hasMessage(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING + "getName");
+    });
+  }
+
+  /**
+   * The test creates an index with a method invocation as part of the expression, changes
+   * the {@link MethodInvocationAuthorizer} to a custom one that still allows the methods part of
+   * the index expression and executes a region operation from the client that would cause an
+   * index mapping removal/addition.
```

Comment:
Could this be reworded to be a little clearer? Maybe something like "changes the {@link MethodInvocationAuthorizer} to a custom one that allows the method that is invoked by the index and executes"

Reply:
Done!.